File: node-js/task4/task4.js

const requestOptions = {
    method: "GET",
    redirect: "follow"
};

const retrieveData = async () => {
    const response = await fetch("https://test-share.shub.edu.vn/api/intern-test/input", requestOptions);
    const data = await response.json();
    return data;
}

const execute = async () => {
    const parsedData = await retrieveData();

    const token = parsedData.token;
    const data = parsedData.data;
    const query = parsedData.query;

    const sum1 = [0]; // sum of all elements
    const sum2 = [0]; // sum of all elements at even index - sum of all elements at odd index

    // Calculate sum1 and sum2 from 0 to i for each i
    for (let i = 0; i < data.length; i++) {
        sum1.push(sum1[i] + data[i]);
        if (i % 2 == 0) {
            sum2.push(sum2[i] + data[i]);
        } else {
            sum2.push(sum2[i] - data[i]);
        }
    }

    const result = [];
    for (let i = 0; i < query.length; i++) {
        if (query[i].type == '1') {
            // if type is 1, calculate sum1 of the range and push to result
            result.push(sum1[query[i].range[1] + 1] - sum1[query[i].range[0]]);
        } else {
            // if type is 2, calculate sum2 of the range and push to result
            if (query[i].range[0] % 2 == 0) {

                result.push(sum2[query[i].range[1] + 1] - sum2[query[i].range[0]]);
            } else {
                // if the first index of the range is odd, the result will be negative
                result.push(sum2[query[i].range[0]] - sum2[query[i].range[1] + 1]);
            }
        }
    }

    await submit(token, result);

}

const submit = async (token, result) => {
    const submitHeaders = new Headers();
    submitHeaders.append("Content-Type", "application/json");
    submitHeaders.append("Authorization", "Bearer " + token);

    const raw = JSON.stringify(result);

    const requestOptions = {
        method: "POST",
        headers: submitHeaders,
        body: raw,
        redirect: "follow"
    };

    fetch("https://test-share.shub.edu.vn/api/intern-test/output", requestOptions)
        .then((response) => response.text())
        .then((result) => console.log(result))
        .catch((error) => console.error(error));
}

execute();

File: golang/task4/task4.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type Query struct {
	Range [2]int `json:"range"`
	Type  string `json:"type"`
}

type ParsedData struct {
	Token string  `json:"token"`
	Data  []int   `json:"data"`
	Query []Query `json:"query"`
}

func main() {
	parsedData, err := fetchData("https://test-share.shub.edu.vn/api/intern-test/input")
	if err != nil {
		fmt.Println(err)
		return
	}

	sum1, sum2 := calculateSums(parsedData.Data)
	result := processQueries(parsedData.Query, sum1, sum2)

	if err := uploadAnswer(parsedData.Token, result); err != nil {
		fmt.Println(err)
	}
}

func calculateSums(data []int) ([]int, []int) {
	sum1, sum2 := []int{0}, []int{0}
	for i, val := range data {
		sum1 = append(sum1, sum1[i]+val)
		if i%2 == 0 {
			sum2 = append(sum2, sum2[i]+val)
		} else {
			sum2 = append(sum2, sum2[i]-val)
		}
	}
	return sum1, sum2
}

func processQueries(queries []Query, sum1, sum2 []int) []int {
	var result []int
	for _, q := range queries {
		if q.Type == "1" {
			result = append(result, sum1[q.Range[1]+1]-sum1[q.Range[0]])
		} else {
			if q.Range[0]%2 == 0 {
				result = append(result, sum2[q.Range[1]+1]-sum2[q.Range[0]])
			} else {
				result = append(result, sum2[q.Range[0]]-sum2[q.Range[1]+1])
			}
		}
	}
	return result
}

func fetchData(url string) (ParsedData, error) {
	res, err := http.Get(url)
	if err != nil {
		return ParsedData{}, err
	}
	defer res.Body.Close()

	body, err := io.ReadAll(res.Body)
	if err != nil {
		return ParsedData{}, err
	}

	var parsedData ParsedData
	if err := json.Unmarshal(body, &parsedData); err != nil {
		return ParsedData{}, err
	}
	return parsedData, nil
}

func uploadAnswer(token string, result []int) error {
	payload, err := json.Marshal(result)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", "https://test-share.shub.edu.vn/api/intern-test/output", strings.NewReader(string(payload)))
	if err != nil {
		return err
	}
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Authorization", "Bearer "+token)

	client := &http.Client{}
	res, err := client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	body, err := io.ReadAll(res.Body)
	if err != nil {
		return err
	}
	fmt.Println(string(body))
	return nil
}
